File: src/SAC/SAC.RaccoltaConsensi/ConsensiPaziente/Site.Master.vb
Imports DI.PortalUser2
Imports DI.PortalUser2.Data
Imports DI.PortalUser2.RoleManager

Partial Public Class Site
    Inherits MasterPage

    Public UtenteCorrente As Utente
    Public ErroreCaricamento As Boolean = False

    Protected ReadOnly Property InstrumentationKey As String
        Get
            Return Microsoft.ApplicationInsights.Extensibility.TelemetryConfiguration.Active.InstrumentationKey
        End Get
    End Property

    Private Sub Page_Init(sender As Object, e As System.EventArgs) Handles Me.Init
        Try
            '
            ' CARICO DA DB LE INFORMAZIONI SULL'UTENTE CORRENTE
            '
            UtenteCorrente = GestioneUtente.GetUtenteCorrente

        Catch ex As Exception
            ErroreCaricamento = True
            Dim sErrorMessage As String = GestioneErrori.TrapError(ex)
            ShowErrorLabel(sErrorMessage)
        End Try
    End Sub

    Protected Sub Page_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load

        If Not IsPostBack Then
            Try

                PageHeader.Visible = Not MySession.IsAccessoDiretto

                FooterPlaceholder.Visible = Not MySession.IsAccessoDiretto

                'Valorizzo la versione del progetto.
                lblVersioneAssembly.Text = Utility.GetAssemblyVersion()

                'Valorizzo il pc.
                lblNomeHost.Text = Utility.GetUserHostName()

                'Nuovo: per ora non lo faccio lascio l'header attuale
                'Dim oUserInterface As UserInterface = New UserInterface(My.Settings.PortalUserConnectionString)
                'HeaderPlaceholder.Text = oUserInterface.GetBootstrapHeader2(PortalsTitles.PortaleConsensi)

                'Popolo le info relative all'utente
                PopolaInfoUtente()

            Catch
                'NON BLOCCO L'ESECUZIONE PER PROBLEMI GRAFICI
            End Try
        End If

    End Sub

    Public Sub ShowErrorLabel(ErrorMessage As String)
        'Utility.ShowErrorLabel(LabelError, ErrorMessage)
        If ErrorMessage.Length > 0 Then
            DivError.Visible = True
            LabelError.Text = ErrorMessage
            LabelError.Visible = True
            LabelError.Style.Add("display", "block")
        End If
    End Sub


    Private Sub PopolaInfoUtente()
        Try
            'Popolo la combo dei ruoli.
            Dim listaRuoli As List(Of Ruolo) = PortalUserSingleton.instance.RoleManagerUtility.GetRuoli()

            'Se l'utente non ha ruoli l'applicazione deve funzionare (non si potrà gestire il consenso generico)
            If (listaRuoli Is Nothing OrElse listaRuoli.Count = 0) Then
                'Nascondo label e comboi dei ruoli
                divRuoli.Visible = False
            Else
                'Popolo la combo dei ruoli
                Call PopolaComboRuoliUtente(listaRuoli)
            End If

            Dim nomeCognomeUtente As String = String.Empty

            'Questo oggetto viene salvato in sessione all'interno del global.asax
            Dim ultimoAccesso As SessioneUtente.UltimoAccesso = CType(Session(Utility.sess_dati_ultimo_accesso), SessioneUtente.UltimoAccesso)
            If (Not ultimoAccesso Is Nothing) Then
                nomeCognomeUtente = $"{ultimoAccesso.Utente?.Nome} {ultimoAccesso.Utente?.Cognome}"
                lblUltimoAccesso.Text = ultimoAccesso.UltimoAccessoDescrizione
            End If

            'valorizzo il nome-cognome dell'utente e il ruolo corrente.
            If (listaRuoli Is Nothing OrElse listaRuoli.Count = 0) Then
                lblInfoUtente.Text = $"{nomeCognomeUtente}"
            Else
                lblInfoUtente.Text = $"{nomeCognomeUtente} - {ddlRuoliUtente.SelectedItem.Text}"
            End If

            'Valorizzo il nome utente e la postazione contenuti nella dropdownlist dei ruoli.
            lblUtente.Text = UtenteCorrente.AccountName
            lblPostazione.Text = Utility.GetUserHostName()

        Catch ex As ApplicationException
            'Nascondo la riga della tabella con la combo dei ruoli.
            divRuoli.Visible = False
            Dim sErrorMessage As String = GestioneErrori.TrapError(ex)
            ShowErrorLabel(sErrorMessage)
            'Utility.NavigateToErrorPage(ErrorPage.ErrorCode.Exception, sErrorMessage, False)

        Catch ex As Exception
            Dim sErrorMessage As String = GestioneErrori.TrapError(ex)
            ShowErrorLabel(sErrorMessage)
            'Utility.NavigateToErrorPage(ErrorPage.ErrorCode.Exception, sErrorMessage, False)
        End Try
    End Sub

    ''' <summary>
    ''' Metodo per popolare la combo dei ruoli
    ''' </summary>
    Private Sub PopolaComboRuoliUtente(listaRuoli As List(Of Ruolo))
        Try
            'ottengo il ruolo corrente
            Dim ruoloCorrenteCodice As String = PortalUserSingleton.instance.RoleManagerUtility.RuoloCorrente.Codice

            'Cancello gli item della combo
            ddlRuoliUtente.Items.Clear()

            'Popolo la combo con i ruoli dell'utente
            ddlRuoliUtente.DataSource = listaRuoli
            ddlRuoliUtente.DataValueField = "Codice"
            ddlRuoliUtente.DataTextField = "Descrizione"
            ddlRuoliUtente.DataBind()

            'Effettuo il bind della combo
            If Not String.IsNullOrEmpty(ruoloCorrenteCodice) Then
                Dim item As ListItem = ddlRuoliUtente.Items.FindByValue(ruoloCorrenteCodice)
                If Not item Is Nothing Then
                    item.Selected = True
                End If
            End If
        Catch ex As Exception
            Dim sErrorMessage As String = GestioneErrori.TrapError(ex)
            ShowErrorLabel(sErrorMessage)
        End Try
    End Sub

    Private Sub ddlRuoliUtente_SelectedIndexChanged(sender As Object, e As EventArgs) Handles ddlRuoliUtente.SelectedIndexChanged
        Try
            Dim currentUserName As String = UtenteCorrente.AccountName
            Dim ddlRuoliUtente As DropDownList = CType(sender, DropDownList)
            'Salvo nella tabella DatiUtente l'ultimo ruolo scelto dall'utente
            PortalUserSingleton.instance.PortalDataAdapterManager.DatiUtenteSalvaValore(currentUserName, RoleManagerUtility2.DI_USER_RUOLO_CORRENTE_CODICE, ddlRuoliUtente.SelectedItem.Value)

            'Scrivo la data di accesso corrente e il ruolo corrente sarà quella letta alla prossima ripartenza della sessione
            PortalUserSingleton.instance.SessioneUtente.SetUltimoAccesso(currentUserName, PortalsNames.PortaleConsensi, DateTime.Now, ddlRuoliUtente.SelectedItem.Value)

            'Imposto endResponse = false per non generarare ThreadAbortException
            Response.Redirect(Request.RawUrl, False)

        Catch ex As Exception
            Dim sErrorMessage As String = GestioneErrori.TrapError(ex)
            ShowErrorLabel(sErrorMessage)
        End Try
    End Sub
End Class

File: src/PortaleUser/DI.OrderEntryPlanner.Site2/Site.Master.cs
using System;
using System.Web;
using System.Web.UI;
using System.Net.NetworkInformation;
using System.Collections;
using DI.PortalUser2;
using System.Web.UI.WebControls;
using Serilog;
using DI.PortalUser2.Data;
using System.Collections.Generic;
using OrderEntryPlanner.Properties;
using DI.PortalUser2.RoleManager;
using OrderEntryPlanner.Components;

namespace OrderEntryPlanner
{
    public partial class SiteMaster : MasterPage
    {
        #region Properties

        private UserInterface userInterface = new UserInterface(Properties.Settings.Default.PortalUserConnectionString);
        private string currentUserName = HttpContext.Current.User.Identity.Name.ToUpper();

        #endregion

        protected void Page_Load(object sender, EventArgs e)
        {
            try
            {
                //Valorizzo il titolo della pagina.
                Page.Title = PortalsTitles.OrderEntryPlanner;

                //Valorizzo la versione del progetto.
                lblVersioneAssembly.Text = Components.Utility.GetAssemblyVersion();

                //Valorizzo il nome del server.
                var oProperties = IPGlobalProperties.GetIPGlobalProperties();
                lblNomeHost.Text = string.Format("{0}.{1}", oProperties.HostName, oProperties.DomainName);

                //Imposto l'header
                HeaderPlaceholder.Text = userInterface.GetBootstrapHeader2(PortalsTitles.OrderEntryPlanner);

                if (!IsPostBack)
                {

                    //Popolo il menu orizzontale
                    PopolaMenuPrincipale();

                    //Popolo le info relative all'utente
                    PopolaInfoUtente();
                }

                //Traccio gli accessi al portale.
                PortalUserTracciaAccessi();
            }
            catch (Exception ex)
            {
                Log.Error(ex, ex.Message);
                showAlert("Si è verificato un errore. Contattare un amministratore di sistema.");
            }
        }

        protected void ddlRuoliUtente_SelectedIndexChanged(object sender, EventArgs e)
        {
            try
            {
                DropDownList ddlRuoliUtente = (DropDownList)(sender);

                //Salvo nella tabella DatiUtente l'ultimo ruolo scelto dall'utente
                PortalUserSingleton.instance.PortalDataAdapterManager.DatiUtenteSalvaValore(currentUserName, RoleManagerUtility2.DI_USER_RUOLO_CORRENTE_CODICE, ddlRuoliUtente.SelectedItem.Value);

                //Scrivo la data di accesso corrente e il ruolo corrente: sarà quella letta alla prossima ripartenza della sessione
                PortalUserSingleton.instance.SessioneUtente.SetUltimoAccesso(currentUserName, PortalsNames.OrderEntryPlanner, DateTime.Now, ddlRuoliUtente.SelectedItem.Value);

                //Imposto endResponse = false per non generarare ThreadAbortException
                Response.Redirect(Request.RawUrl, false);
            }
            catch (Exception ex)
            {
                Log.Error(ex, ex.Message);
                showAlert("Si è verificato un errore. Contattare un amministratore di sistema.");
            }
        }

        #region FunzioniUtility
        /// <summary>
        /// Metodo che popola il menu principale (cross-application)
        /// </summary>
        private void PopolaMenuPrincipale()
        {
            try
            {
                //Elimino tutti gli elementi dal menu.
                MenuMain.Items.Clear();

                //Ottengo l'id del portale.
                int idPortale = (int)PortalDataAdapterManager.EnumPortalId.OrderEntryPlanner;

                //Ottengo le voci del menu.
                List<PortalUserMenuItem> listaPortalUserMenuItem = PortalUserSingleton.instance.PortalDataAdapterManager.GetMainMenu(idPortale);

                //Aggiungo le voci del menu.
                foreach (var menuItem in listaPortalUserMenuItem)
                {
                    MenuItem newMenuItem = new MenuItem(menuItem.Descrizione, null, null, ResolveUrl(menuItem.Url), string.Empty);
                    MenuMain.Items.Add(newMenuItem);
                    newMenuItem.Selected = menuItem.IsSelected;
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex, ex.Message);
                showAlert("Si è verificato un errore. Contattare un amministratore di sistema.");
            }
        }

        /// <summary>
        /// Funzione utilizzata per visualizzare il messaggio di errore.
        /// </summary>
        /// <param name="text"></param>
        public void showAlert(string text)
        {
            try
            {
                DivError.Visible = text.Length > 0;
                DivError.InnerText = text;

                if (text.Length > 0) DivError.Style.Remove("display");
                else DivError.Style.Add("display", "none");
                updError.Update();

            }
            catch (Exception ex)
            {
                Log.Error(ex, ex.Message);
                showAlert("Si è verificato un errore. Contattare un amministratore di sistema.");
            }
        }


        /// <summary>
        /// Metodo che popola le info relative all'utente
        /// </summary>
        private void PopolaInfoUtente()
        {
            try
            {
                //Popolo la combo dei ruoli.
                List<Ruolo> listaRuoli = PortalUserSingleton.instance.RoleManagerUtility.GetRuoli();

                //Se l'utente non ha ruoli allora restituisco un errore.
                if (listaRuoli == null || listaRuoli.Count == 0)
                {
                    throw new ApplicationException("L'utente corrente non ha ruoli associati.");
                }

                //Popolo la combo dei ruoli
                PopolaComboRuoliUtente(listaRuoli);

                string nomeCognomeUtente = string.Empty;

                //Questo oggetto viene salvato in sessione all'interno del global.asax
                SessioneUtente.UltimoAccesso ultimoAccesso = (SessioneUtente.UltimoAccesso)Session[Components.Utility.sess_dati_ultimo_accesso];
                if (ultimoAccesso != null)
                {
                    nomeCognomeUtente = $"{ultimoAccesso.Utente?.Nome} {ultimoAccesso.Utente?.Cognome}";
                    lblUltimoAccesso.Text = ultimoAccesso.UltimoAccessoDescrizione;
                }

                //valorizzo il nome-cognome dell'utente e il ruolo corrente.
                lblInfoUtente.Text = $"{nomeCognomeUtente} - {ddlRuoliUtente.SelectedItem.Text}";

                //Valorizzo il nome utente e la postazione contenuti nella dropdownlist dei ruoli.
                lblUtente.Text = currentUserName;
                lblPostazione.Text = Components.Utility.GetUserHostName();
            }
            catch (ApplicationException ex)
            {
                //Nascondo la riga della tabella con la combo dei ruoli.
                divRuoli.Visible = false; //TODO: @SimoneB - Verificare se serve.
                Log.Error(ex, ex.Message);
                Components.Utility.NavigateToErrorPage(Components.Utility.ErrorCode.Exception, ex.Message, false);
            }
            catch (Exception ex)
            {
                Log.Error(ex, ex.Message);
                showAlert("Si è verificato un errore. Contattare un amministratore di sistema.");
            }
        }

        /// <summary>
        ///  Metodo per popolare la combo dei ruoli
        /// </summary>
        private void PopolaComboRuoliUtente(List<Ruolo> listaRuoli)
        {
            try
            {
                //ottengo il ruolo corrente
                string ruoloCorrenteCodice = PortalUserSingleton.instance.RoleManagerUtility.RuoloCorrente.Codice;

                //Cancello gli item della combo
                ddlRuoliUtente.Items.Clear();

                //Popolo la combo con i ruoli dell'utente
                ddlRuoliUtente.DataSource = listaRuoli;
                ddlRuoliUtente.DataValueField = "Codice";
                ddlRuoliUtente.DataTextField = "Descrizione";
                ddlRuoliUtente.DataBind();

                //Effettuo il bind della combo
                if (ruoloCorrenteCodice != null)
                {
                    ListItem item = ddlRuoliUtente.Items.FindByValue(ruoloCorrenteCodice);
                    if (item != null)
                    {
                        item.Selected = true;
                    }
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex, ex.Message);
                showAlert("Si è verificato un errore. Contattare un amministratore di sistema.");
            }
        }

        /// <summary>
        /// Funzione che traccia gli accessi al PortalUser
        /// </summary>
        private void PortalUserTracciaAccessi()
        {
            try
            {
                //Arrivato a questo punto scrivo in sessione i dati di accesso: lo faccio una volta sola a sessione
                if (Session[Components.Utility.sess_write_dati_accesso] == null)
                {
                    //In caso di errore dovuto alla mancanza di ruoli per l'utente cmbRuoliUtente.SelectedItem è null
                    if (ddlRuoliUtente.SelectedItem != null)
                    {
                        //Setto a true la relativa variabile di sessione
                        Session[Components.Utility.sess_write_dati_accesso] = true;

                        //Scrivo la data e il ruolo dell'accesso corrente. La prossima sessione ripartirà da questi elementi
                        PortalUserSingleton.instance.SessioneUtente.SetUltimoAccesso(currentUserName, PortalsNames.OrderEntryPlanner, DateTime.Now, ddlRuoliUtente.SelectedItem.Value);

                        //Traccio l'accesso al portale.
                        string msgTracciamentoAccessi = $"Accesso effettuato il {DateTime.Now.ToString("dd/MM/yyyy")} alle ore {DateTime.Now.ToString("HH:mm:ss")}";
                        PortalUserSingleton.instance.PortalDataAdapterManager.TracciaAccessi(HttpContext.Current.User.Identity.Name, PortalsNames.OrderEntryPlanner, msgTracciamentoAccessi, ddlRuoliUtente.SelectedValue);
                    }
                }
            }
            catch (Exception ex)
            {
                Log.Error(ex, ex.Message);
                showAlert("Si è verificato un errore. Contattare un amministratore di sistema.");
            }
        }
        #endregion
    }
}